### PR TITLE
add `abstrct` dataset

### PR DESCRIPTION
**Pull request by ArneBinder**
In this PR, we add AbsTRCT - an argumentation-mining dataset, according to the description in https://github.com/ArneBinder/pie-datasets/issues/10.

### CODE REVIEW

**Review by ArneBinder — COMMENTED** [review]
Looks very good, just some minor points, see below.

File: dataset_builders/pie/abstrct/abstrct.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "neoplasm_dev": "abstrct-master/AbstRCT_corpus/data/dev/neoplasm_dev",
+    "neoplasm_test": "abstrct-master/AbstRCT_corpus/data/test/neoplasm_test",
+    "glaucoma_test": "abstrct-master/AbstRCT_corpus/data/test/glaucoma_test",
+    "mixed_test": "abstrct-master/AbstRCT_corpus/data/test/mixed_test",
+}
+
+
+class AbstRCT(BratBuilder):
+    BASE_DATASET_PATH = "DFKI-SLT/brat"
+    BASE_DATASET_REVISION = "844de61e8a00dc6a93fc29dc185f6e617131fbf1"
```
[the remark this answers is not in the document]
Comment (ArneBinder) [reply dataset_builders/pie/abstrct/abstrct.py]: I fixed this in 74cbfd9a47101a98e1995b09a377907dd67fcd36

File: tests/dataset_builders/pie/test_abstrct.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+from pie_datasets import DatasetDict
+from pie_datasets.builders.brat import BratDocument, BratDocumentWithMergedSpans
+from tests.dataset_builders.common import (
+    PIE_BASE_PATH,
+    TestTokenDocumentWithLabeledSpansAndBinaryRelations,
+)
+
+disable_caching()
+
+DATASET_NAME = "abstrct"
```
[the remark this answers is not in the document]
Comment (idalr) [reply tests/dataset_builders/pie/test_abstrct.py:19]: changed in 16f0152

File: tests/dataset_builders/pie/test_abstrct.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    else:
+        raise ValueError(f"Unknown dataset variant: {dataset_variant}")
+    return result
+
+
+def test_document(document, dataset_variant):
+    assert document is not None
+    assert document.text.startswith(
+        " A combination of mitoxantrone plus prednisone is preferable to prednisone alone"
+    )
```
[the remark this answers is not in the document]
Comment (idalr) [reply tests/dataset_builders/pie/test_abstrct.py:70]: as discussed, I made changes in c6fd863

File: tests/dataset_builders/pie/test_abstrct.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    document_converters = builder.document_converters
+
+    if dataset_variant == "default":
+        assert document_converters == {}
+    elif dataset_variant == "merge_fragmented_spans":
+        assert len(document_converters) == 1
+        assert set(document_converters) == {
+            TextDocumentWithLabeledSpansAndBinaryRelations,
+        }
+        assert all(dict(v) for k, v in document_converters.items())
```
[the remark this answers is not in the document]
Comment (idalr) [reply tests/dataset_builders/pie/test_abstrct.py]: changed in c6fd863

File: tests/dataset_builders/pie/test_abstrct.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArneBinder) [reply tests/dataset_builders/pie/test_abstrct.py:70]: please do not rely on the metadata, just check the actual annotations (of the `spans` layer)

[the remark this answers is not in the document]
Comment (idalr) [reply tests/dataset_builders/pie/test_abstrct.py:87]: edited in [ac1c097].
ps. not sure if L66 is a good implementation, but couldn't work as in L790

File: tests/dataset_builders/pie/test_abstrct.py [hunk 5]
```diff
@@ -60,40 +60,51 @@ def test_document(document, dataset_variant):
     assert document is not None
     assert document.id == "10561201"
 
+    # check the annotation
+    if dataset_variant == "default":
+        span_texts_labels_tuples = [
+            (document.text[span.slices[0][0] : span.slices[-1][1]], span.label)
+            for span in document.spans
+        ]
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:68]:
```suggestion
        span_texts_labels_tuples = [
            (tuple(document.text[start:end] for start, end in span.slices), span.label)
            for span in document.spans
        ]
```

File: tests/dataset_builders/pie/test_abstrct.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArneBinder) [reply tests/dataset_builders/pie/test_abstrct.py:87]: see below: https://github.com/ArneBinder/pie-datasets/pull/68#discussion_r1430660109

**Comment by idalr** [comment]
While the AbstRCT dataset has no fragmented spans, we converted the dataset to a single-variant version ("default") in 612466b.

File: dataset_builders/pie/abstrct/README.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+doc = datasets["neoplasm_train"][0]
+assert isinstance(doc, builders.brat.BratDocumentWithMergedSpans)
+```
+
+### Document Converters
+
+The dataset provides document converters for the following target document types:
+
+- `pytorch_ie.documents.TextDocumentWithLabeledSpansAndBinaryRelations`
+  - `LabeledSpans`, converted from `BratDocument`'s `spans`
```
Comment (ArneBinder) [inline dataset_builders/pie/abstrct/README.md:49]:
```suggestion
  - `LabeledSpans`, converted from `BratDocumentWithMergedSpans`'s `spans`
```

File: dataset_builders/pie/abstrct/README.md [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    - labels: `MajorClaim`, `Claim`, `Premise`
+  - `BinraryRelations`, converted from `BratDocument`'s `relations`
```
Comment (ArneBinder) [inline dataset_builders/pie/abstrct/README.md:51]:
```suggestion
  - `BinraryRelations`, converted from `BratDocumentWithMergedSpans`'s `relations`
```

File: dataset_builders/pie/abstrct/abstrct.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        BratConfig(name=BratBuilder.DEFAULT_CONFIG_NAME, merge_fragmented_spans=True),
+    ]
+    DOCUMENT_TYPES = {
+        BratBuilder.DEFAULT_CONFIG_NAME: BratDocumentWithMergedSpans,
+    }
+
+    # we need to add None to the list of dataset variants to support the default dataset variant
+    BASE_BUILDER_KWARGS_DICT = {
+        dataset_variant: {"url": URL, "split_paths": SPLIT_PATHS}
+        for dataset_variant in ["default", "merge_fragmented_spans", None]
```
Comment (ArneBinder) [inline dataset_builders/pie/abstrct/abstrct.py:30]:
```suggestion
        for dataset_variant in ["default", None]
```

File: tests/dataset_builders/pie/test_abstrct.py [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    return result
+
+
+def test_document(document, dataset_variant):
+    assert document is not None
+    assert document.id == "10561201"
+
+    # check the annotation
+    if dataset_variant == "default" or dataset_variant is None:
+        span_texts_labels_tuples = [(str(span), span.label) for span in document.spans]
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:76]:
```suggestion
    span_texts_labels_tuples = [(str(span), span.label) for span in document.spans]
```

File: tests/dataset_builders/pie/test_abstrct.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArneBinder) [reply tests/dataset_builders/pie/test_abstrct.py:93]: this is not relevant anymore, see https://github.com/ArneBinder/pie-datasets/pull/68#issuecomment-1863094780

File: tests/dataset_builders/pie/test_abstrct.py [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+    document.relations[4].tail == document.spans[6]
+    document.relations[5].label == "Support"
+    document.relations[5].head == document.spans[5]
+    document.relations[5].tail == document.spans[0]
+
+
+@pytest.fixture(scope="module")
+def dataset_of_text_documents_with_labeled_spans_and_binary_relations(
+    dataset, dataset_variant
+) -> Optional[DatasetDict]:
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:141]:
```suggestion
) -> DatasetDict:
```

File: tests/dataset_builders/pie/test_abstrct.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+        )
+    else:
+        raise ValueError(f"Unknown dataset variant: {dataset_variant}")
+    return converted_dataset
+
+
+def test_dataset_of_text_documents_with_labeled_spans_and_binary_relations(
+    dataset_of_text_documents_with_labeled_spans_and_binary_relations,
+):
+    if dataset_of_text_documents_with_labeled_spans_and_binary_relations is not None:
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:154]: remove

File: tests/dataset_builders/pie/test_abstrct.py [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+def tokenizer() -> PreTrainedTokenizer:
+    return AutoTokenizer.from_pretrained("bert-base-uncased")
+
+
+@pytest.fixture(scope="module")
+def tokenized_documents_with_labeled_spans_and_binary_relations(
+    dataset_of_text_documents_with_labeled_spans_and_binary_relations, tokenizer
+) -> Optional[List[TestTokenDocumentWithLabeledSpansAndBinaryRelations]]:
+    if dataset_of_text_documents_with_labeled_spans_and_binary_relations is None:
+        return None
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:258]: remove

File: tests/dataset_builders/pie/test_abstrct.py [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+
+@pytest.fixture(scope="module")
+def tokenizer() -> PreTrainedTokenizer:
+    return AutoTokenizer.from_pretrained("bert-base-uncased")
+
+
+@pytest.fixture(scope="module")
+def tokenized_documents_with_labeled_spans_and_binary_relations(
+    dataset_of_text_documents_with_labeled_spans_and_binary_relations, tokenizer
+) -> Optional[List[TestTokenDocumentWithLabeledSpansAndBinaryRelations]]:
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:256]:
```suggestion
) -> List[TestTokenDocumentWithLabeledSpansAndBinaryRelations]:
```

File: tests/dataset_builders/pie/test_abstrct.py [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+        strict_span_conversion=True,
+        verbose=True,
+    )
+    return tokenized_docs
+
+
+def test_tokenized_documents_with_labeled_spans_and_binary_relations(
+    tokenized_documents_with_labeled_spans_and_binary_relations,
+):
+    if tokenized_documents_with_labeled_spans_and_binary_relations is not None:
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:278]: remove

File: tests/dataset_builders/pie/test_abstrct.py [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+            "'associated', 'with', 'greater', 'and', 'longer', '-', 'lasting', 'improvement', 'in', 'several', "
+            "'hq', '##l', 'domains', 'and', 'symptoms', 'than', 'treatment', 'with', 'pre', '##d', '##nis', '##one', "
+            "'alone', '.')"
+        )
+
+
+def test_tokenized_documents_with_entities_and_relations_all(
+    dataset_of_text_documents_with_labeled_spans_and_binary_relations, tokenizer, dataset_variant
+):
+    if dataset_of_text_documents_with_labeled_spans_and_binary_relations is not None:
```
Comment (ArneBinder) [inline tests/dataset_builders/pie/test_abstrct.py:342]: remove

**Review by ArneBinder — COMMENTED** [review]
Looks very good, just some minors.

File: tests/dataset_builders/pie/test_abstrct.py (hunk 5 above)
Comment (ArneBinder) [reply tests/dataset_builders/pie/test_abstrct.py]: this is outdated
